src/gui/scheduling/syncscheduler.cpp

```diff
@@ -0,0 +1,164 @@
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "gui/scheduling/syncscheduler.h"
+
+#include "gui/folderman.h"
+#include "gui/scheduling/etagwatcher.h"
+#include "libsync/configfile.h"
+
+using namespace std::chrono_literals;
+
+using namespace OCC;
+
+
+class FolderPriorityQueue
+{
+public:
+    FolderPriorityQueue() = default;
+
+    void enqueueFolder(Folder *folder, SyncScheduler::Priority priority)
+    {
+        auto old = _scheduledFolders.find(folder);
+        if (old == _scheduledFolders.cend()) {
+            // the folder is not yet scheduled
+            _queue.emplace(folder, priority);
+        } else {
+            _scheduledFolders.insert({folder, priority});
+            // if the new priority is higher we need to rebuild the queue
+            if (priority > old->second) {
+                // we need to reorder the queue
+                // this is expensive
+                decltype(_queue) out;
+                for (; !_queue.empty(); _queue.pop()) {
+                    const auto &tmp = _queue.top();
+                    if (tmp.folder != folder) {
+                        out.push(std::move(tmp));
+                    } else {
+                        out.emplace(folder, priority);
+                    }
+                }
+                _queue = std::move(out);
+            }
+        }
+    }
+
+    bool empty() { return _queue.empty(); }
+
+    QPointer<Folder> pop()
+    {
+        QPointer<Folder> out;
+        while (!_queue.empty() && !out) {
+            // could be a nullptr by now
+            out = _queue.top().folder;
+            _scheduledFolders.erase(_queue.top().rawFolder);
+            _queue.pop();
+        }
+        return out;
+    }
+
+private:
+    struct Element
+    {
+        Element(Folder *f, SyncScheduler::Priority p)
+            : folder(f)
+            , rawFolder(f)
+            , priority(p)
+        {
+        }
+
+        // We don't own the folder, so it might get deleted
+        QPointer<Folder> folder;
+        // raw pointer for lookup in _scheduledFolders
+        Folder *rawFolder;
+        SyncScheduler::Priority priority;
+
+        friend bool operator<(const Element &lhs, const Element &rhs) { return lhs.priority < rhs.priority; }
+    };
+
+    // the actual queue
+    std::priority_queue<Element> _queue;
+    // helper container to ensure we don't enqueue a Folder multiple times
+    std::unordered_map<Folder *, SyncScheduler::Priority> _scheduledFolders;
```

Comment:
Might be more efficient to store a reference to the element in the queue as the value instead of the raw priority. Querying the priority is still possible, but it could speed up the find calls when looking for the duplicate in the priority queue.

Reply:
Are you sure this is possible without making assumptions about the implementation of priority_queue

Reply:
Also this is not  a hot path.